File: frontend/src/lib/api/text-analysis.ts
import api from './client'
import { downloadFromApi } from './download'

// Text Analysis types
export interface SubgroupFilter {
  column_id: number
  operator: string
  values?: string[]
  value?: string
}

export interface CodeFrequencyBrief {
  code_id: number
  code_name: string
  code_color: string | null
  count: number
  percentage: number
}

export interface FrequencySet {
  row_count: number
  text_count: number
  frequencies: CodeFrequencyBrief[]
}

export interface FilteredFrequenciesResponse {
  filtered: FrequencySet
  overall: FrequencySet | null
  filter_description: string
  filter_scope: { filtered_datasets: string[]; unfiltered_datasets: string[] }
}

export interface CrossTabRow {
  code_id: number
  code_name: string
  code_color: string | null
  counts: Record<string, number>
  percentages: Record<string, number>
  row_total: number
}

export interface CrossTabulationResponse {
  cross_column_name: string
  response_values: string[]
  matrix: CrossTabRow[]
  column_totals: Record<string, number>
  total_coded_texts: number
}

export interface CodeDensityGroup {
  group_value: string
  avg_codes_per_text: number
  text_count: number
}

export interface CodeDensityResponse {
  groups: CodeDensityGroup[]
  overall: CodeDensityGroup
}

export interface ResponseLengthCode {
  code_id: number
  code_name: string
  code_color: string | null
  avg_words: number
  text_count: number
}

export interface ResponseLengthResponse {
  codes: ResponseLengthCode[]
  uncoded: { avg_words: number; text_count: number }
}

// API functions - Text Analysis
// Track J · J2 slab 3b — `layer_scope` ('human' default | 'consensus') selects the coder layer.
export type LayerScope = 'human' | 'consensus'

export const textAnalysisApi = {
  filteredFrequencies: (pid: number, data: { column_ids: number[]; filters: SubgroupFilter[]; include_overall: boolean; coder_ids?: number[] | null; layer_scope?: LayerScope }) =>
    api.post<FilteredFrequenciesResponse>(`/projects/${pid}/text-analysis/filtered-frequencies`, data).then(r => r.data),

  crossTabulation: (pid: number, data: { text_column_ids: number[]; cross_column_id: number; code_ids?: number[]; coder_ids?: number[] | null; layer_scope?: LayerScope }) =>
    api.post<CrossTabulationResponse>(`/projects/${pid}/text-analysis/cross-tabulation`, data).then(r => r.data),

  codeDensity: (pid: number, params: { column_ids: string; group_by_column_id?: number; coder_ids?: string; layer_scope?: LayerScope }) =>
    api.get<CodeDensityResponse>(`/projects/${pid}/text-analysis/code-density`, { params }).then(r => r.data),

  responseLength: (pid: number, params: { column_ids: string; coder_ids?: string; layer_scope?: LayerScope }) =>
    api.get<ResponseLengthResponse>(`/projects/${pid}/text-analysis/response-length-by-code`, { params }).then(r => r.data),

  exportCrossAnalysis: (pid: number, params: { column_ids: string; filters?: string; cross_column_id?: number; coder_ids?: string; layer_scope?: LayerScope }) => {
    const searchParams = new URLSearchParams()
    searchParams.append('column_ids', params.column_ids)
    if (params.filters) searchParams.append('filters', params.filters)
    if (params.cross_column_id) searchParams.append('cross_column_id', String(params.cross_column_id))
    if (params.coder_ids) searchParams.append('coder_ids', params.coder_ids)
    if (params.layer_scope) searchParams.append('layer_scope', params.layer_scope)
    const qs = searchParams.toString()
    return downloadFromApi(`/projects/${pid}/text-analysis/export?${qs}`, 'text-analysis.csv')
  },
}

File: backend/app/schemas/text_analysis.py
"""Pydantic schemas for comment cross-analysis endpoints."""

from pydantic import BaseModel


# ── Request schemas ──────────────────────────────────────────────────────────

class SubgroupFilter(BaseModel):
    """A single filter criterion for subgroup analysis."""
    column_id: int
    operator: str  # "equals", "in", "gte", "lte", "above_mean", "below_mean"
    values: list[str] | None = None  # for equals/in
    value: float | None = None  # for gte/lte


class FilteredFrequenciesRequest(BaseModel):
    column_ids: list[int]
    filters: list[SubgroupFilter] = []
    include_overall: bool = True
    # Track J · J1 item 4 — scope analysis output to selected coders (None/empty = all coders)
    coder_ids: list[int] | None = None
    # Track J · J2 slab 3b — coder layer: None/'human' (default) or 'consensus'
    layer_scope: str | None = None


class CrossTabulationRequest(BaseModel):
    text_column_ids: list[int]
    cross_column_id: int
    code_ids: list[int] | None = None
    # Track J · J1 item 4 — scope analysis output to selected coders (None/empty = all coders)
    coder_ids: list[int] | None = None
    # Track J · J2 slab 3b — coder layer: None/'human' (default) or 'consensus'
    layer_scope: str | None = None


# ── Response schemas ─────────────────────────────────────────────────────────

class CodeFrequencyBrief(BaseModel):
    code_id: int
    code_name: str
    code_color: str | None
    count: int
    percentage: float


class FrequencySet(BaseModel):
    row_count: int
    text_count: int
    frequencies: list[CodeFrequencyBrief]


class FilteredFrequenciesResponse(BaseModel):
    filtered: FrequencySet
    overall: FrequencySet | None = None
    filter_description: str
    filter_scope: dict  # {filtered_datasets: [...], unfiltered_datasets: [...]}


class CrossTabRow(BaseModel):
    code_id: int
    code_name: str
    code_color: str | None
    counts: dict[str, int]  # response_value -> count
    percentages: dict[str, float]  # response_value -> percentage
    row_total: int


class CrossTabulationResponse(BaseModel):
    cross_column_name: str
    response_values: list[str]
    matrix: list[CrossTabRow]
    column_totals: dict[str, int]
    total_coded_texts: int


class CodeDensityGroup(BaseModel):
    group_value: str
    avg_codes_per_text: float
    text_count: int


class CodeDensityResponse(BaseModel):
    groups: list[CodeDensityGroup]
    overall: CodeDensityGroup


class ResponseLengthCode(BaseModel):
    code_id: int
    code_name: str
    code_color: str | None
    avg_words: float
    text_count: int


class ResponseLengthUncoded(BaseModel):
    avg_words: float
    text_count: int


class ResponseLengthResponse(BaseModel):
    codes: list[ResponseLengthCode]
    uncoded: ResponseLengthUncoded
